lib/fiddle.rb

```diff
@@ -1,6 +1,13 @@
 # frozen_string_literal: true
 
-require 'fiddle.so'
+if RUBY_ENGINE == "ruby"
+  require 'fiddle.so'
+else
+  $LOAD_PATH.delete(__dir__)
```

Comment:
Why do we need this?

Reply:
Because we need to load fiddle.rb from the stdlib in TruffleRuby/JRuby, which works, unlike the version of Fiddle in this repo which only works on CRuby.

Reply:
I don't think that changing `$LOAD_PATH` isn't a good approach because `$LOAD_PATH` is one of global variables.
Can we use another approach?

Reply:
There is no other way (AFAIK). We need to make sure the Ruby files of this gem are not loaded because it's a different version than in stdlib. So we need to remove it from $LOAD_PATH.
It is the same approach as for older FFI versions: https://github.com/ffi/ffi/blob/85d0fabce6ab5ceb3848f7e09a265341672a272d/lib/ffi.rb#L19

Reply:
I feel that it's an ad-hoc approach.
Can we use https://github.com/oracle/truffleruby/blob/512427a658f29b222df9a6da1b3ffc12b46a3be6/lib/mri/fiddle.rb#L3-L4 ?

Reply:
> I feel that it's an ad-hoc approach.

It is the only approach that reliably works (AFAIK).

> Can we use https://github.com/oracle/truffleruby/blob/512427a658f29b222df9a6da1b3ffc12b46a3be6/lib/mri/fiddle.rb#L3-L4 ?

No, that wouldn't work because then we would mix this gem's version of the Ruby files with the backend files expecting a different Fiddle version.
As an example if any method moved between the Ruby files and the C extension here it would break. Or if the Ruby files expect some method to be defined in the C extension that would also break.
In addition since TruffleRuby is not yet in this repo's CI it would be very likely to break silently.

When TruffleRuby passes all/almost-all Fiddle tests we will consider upstreaming what makes sense and analyze what's the best interface (I'm thinking either upstream the pure-Ruby backend or use the C extension's methods as the interface but then it needs to be more stable and won't be very practical. or define a separate interface between Fiddle frontend & backend). It is too early for that since many tests don't pass.

Reply:
> > I feel that it's an ad-hoc approach.
> 
> It is the only approach that reliably works (AFAIK).

This approach depends on `$LOAD_PATH` contents. So you want to add `test/run.rb` changes, right?

Can we prepare an empty fiddle gem only for TruffleRuby?

Reply:
> This approach depends on `$LOAD_PATH` contents.

Yes, there is no other way for `require 'fiddle'` anywhere in the program to load from stdlib and not from the gem (AFAIK).

> So you want to add `test/run.rb` changes, right?

That file is adding "absolute" non-normalized paths to `$LOAD_PATH`, so it's just fixing that. Applications typically don't have this issue as the `$LOAD_PATH` entry for the fiddle gem will be added by RubyGems or Bundler.

> Can we prepare an empty fiddle gem only for TruffleRuby?

There is no TruffleRuby-specific `Gem::Platform` (https://github.com/rubygems/rubygems/blob/master/lib/rubygems/platform.rb).
Also this issue of fiddle becoming a bundled gem affects multiple Ruby implementations, so it's not just about TruffleRuby but also JRuby and maybe others.
(and it's also not just about fiddle, it's the same for all default->bundled gems transitions when the gem has a C extension and the extension doesn't just work on these Rubies).
Reusing the stdlib is the obvious short-term fix because that's what was used before and known to be working & tested in that Ruby implementation.
In short, it unbreaks what the warnings in https://bugs.ruby-lang.org/issues/20309 broke.
Longer-term these Ruby implementations might be able to upstream their Fiddle backend in this repo.

Reply:
Can we prepend the system path to `$LOAD_PATH` instead of deleting a path from `$LOAD_PATH` instead something like the following?

```ruby
system_fiddle_lib_dir = RbConfig::CONFIG["rubylibdir"]
begin
  load_path = $LOAD_PATH.dup
  $LOAD_PATH.prepend(system_fiddle_lib_dir)
  require "fiddle"
ensure
  # The above 'require "fiddle"' must not change $LOAD_PATH but...
  $LOAD_PATH.shift if $LOAD_PATH[0] == system_fiddle_lib_dir
end
```

Reply:
Can we use `require "#{system_fiddle_lib_dir}/fiddle"` and `require_relative` in TruffleRuby's Fiddle?

Reply:
Both of these are far more complicated (and are likely broken for some cases, e.g. `require "fiddle"` might repetitively change $LOAD_PATH for the former) than the change in this PR.
What's the concrete problem with that change? It's simple and it does the job.

Reply:
> Can we use `require "#{system_fiddle_lib_dir}/fiddle"` and `require_relative` in TruffleRuby's Fiddle?

It wouldn't work if user code does e.g. `require 'fiddle/closure'`.
And changing fiddle files in TruffleRuby is increasing the diff + it does not fix the issue for existing TruffleRuby releases.

Reply:
> What's the concrete problem with that change?

I don't want to change a global variable as I already said.

> It wouldn't work if user code does e.g. `require 'fiddle/closure'`.

`require 'fiddle/closure'` doesn't work with your `$LOAD_PATH.delete` approach.

BTW, why do users need to use `require "fiddle/closure"`? `fiddle.rb` requires `fiddle/closure.rb` automatically.

> changing fiddle files in TruffleRuby is increasing the diff to upstream

ruby/fiddle also uses `require_relative`.

> it does not fix the issue for existing TruffleRuby releases.

Can users pin old gems that don't have explicit `fiddle` dependencies for the case?

Reply:
> I don't want to change a global variable as I already said.

It's not Fiddle's global variable.
`$LOAD_PATH` is the only way to change what `require "fiddle"` finds (AFAIK), so it's like part of the "API" for `require`.
We are basically undoing what RubyGems did by adding the Fiddle gem to `$LOAD_PATH`, which is exactly what we want here as the only thing that works is using the stdlib.

> `require 'fiddle/closure'` doesn't work with your `$LOAD_PATH.delete` approach.

It does if the order is `require 'fiddle'` and then `require 'fiddle/closure'` (in user code).
But indeed it doesn't if it's `require 'fiddle/closure'` (in user code) before `require 'fiddle'`.

> BTW, why do users need to use `require "fiddle/closure"`? `fiddle.rb` requires `fiddle/closure.rb` automatically.

Agreed, they shouldn't. It illustrates that changing `$LOAD_PATH` just for the duration of that require is brittle.

> ruby/fiddle also uses `require_relative`.

No it doesn't:
https://github.com/ruby/fiddle/blob/aad5a3bc797c7271bbc06258fe5cd99cd3f3d2bb/lib/fiddle.rb#L4-L6

> > it does not fix the issue for existing TruffleRuby releases.
> 
> Can users pin old gems that don't have explicit `fiddle` dependencies for the case?

That seems too complicated and impractical. Many things in the Ruby ecosystem automatically update to the latest gem versions, etc. Also the idea is an application running on CRuby should run on TruffleRuby as-is, without changing the Gemfile (because doing so might then bring back some bugs in gems, etc). Having to figure out this kind of version constraints is a nightmare and too impractical.

Reply:
@hsbt In https://bugs.ruby-lang.org/issues/18567#note-35 you said:
> I can help default/bundled gems are working with JRuby/TruffleRuby like:

I would like to ask your help here in convincing @kou this is the way to go, or to find an alternative if you know one.
I don't see any other solution which would work on existing TruffleRuby & JRuby releases.
https://bugs.ruby-lang.org/issues/20309 is breaking TruffleRuby & JRuby, this PR is AFAIK the only short-term solution.
This is BTW nothing new, it's the approach I mentioned in https://bugs.ruby-lang.org/issues/18567#note-30 (for strscan we eventually found an approach of not shipping a `strscan.rb` in the gem, but that's not an option here).

Reply:
> It's not Fiddle's global variable.

There is no "XXX's global variable". Because it's "global".
I said that I don't want to change any global variables by `require "fiddle"`.

> `$LOAD_PATH` is the only way to change what `require "fiddle"` finds (AFAIK)

`require "#{system_fiddle_lib_dir}/fiddle" and `require_relative` can change it as I already mentioned.

> > BTW, why do users need to use `require "fiddle/closure"`? `fiddle.rb` requires `fiddle/closure.rb` automatically.
> 
> Agreed, they shouldn't. It illustrates that changing `$LOAD_PATH` just for the duration of that require is brittle.

We don't need to care about unexpected cases.
It shows that the approach doesn't have real world problem.

> > ruby/fiddle also uses `require_relative`.
> 
> No it doesn't:

Ah, sorry. I wanted to say that we can change the current `require` to `require_relative` in ruby/fiddle. So "increasing the diff to upstream" is not a blocker.

> > > it does not fix the issue for existing TruffleRuby releases.
> > 
> > 
> > Can users pin old gems that don't have explicit `fiddle` dependencies for the case?
> 
> That seems too complicated and impractical. Many things in the Ruby ecosystem automatically update to the latest gem versions, etc. Also the idea is an application running on CRuby should run on TruffleRuby as-is, without changing the Gemfile (because doing so might then bring back some bugs in gems, etc). Having to figure out this kind of version constraints is a nightmare and too impractical.

If TruffleRuby works on the real fix ("Longer-term these Ruby implementations might be able to upstream their Fiddle backend in this repo") as soon as possible, we can avoid this situation.

If this is a really short-term fix, we don't need `require_relative` changes. Because `lib/fiddle/**/*.rb` in ruby/fiddle and TruffleRuby will not be different. So existing TruffleRuby releases will work with `lib/fiddle/**/*.rb` in ruby/fiddle.

Reply:
> I said that I don't want to change any global variables by require "fiddle".

That wasn't clear to me, and I guess you mean change any global variable permanently, and that's why you prefer https://github.com/ruby/fiddle/pull/146#discussion_r1772483803 to just `$LOAD_PATH.delete(__dir__)`, correct?

I think we can try that solution, I will need to check it behaves correctly for multiple `require 'fiddle'` calls.

> I wanted to say that we can change the current require to require_relative in ruby/fiddle. So "increasing the diff to upstream" is not a blocker.

Right, it's fine to change it here, but it doesn't solve the issue for existing truffleruby & jruby releases.

> If TruffleRuby works on the real fix ("Longer-term these Ruby implementations might be able to upstream their Fiddle backend in this repo") as soon as possible, we can avoid this situation.

There are many such gems which became bundled gems, so it's not an issue only for fiddle but for [a number of gems](https://github.com/oracle/truffleruby/issues/2644). For some of these gems (but not all) we can just reuse the C extension on TruffleRuby though, so for those gems it's then only a problem for JRuby.

In terms of priority for upstreaming TruffleRuby backend into the respective gems, `strscan` feels higher than `fiddle` because there seem to be more frequent changes there.
Anyway, we will get to it, but to solve it for existing releases we need this PR.

> If this is a really short-term fix, we don't need require_relative changes. Because `lib/fiddle/**/*.rb` in ruby/fiddle and TruffleRuby will not be different. So existing TruffleRuby releases will work with `lib/fiddle/**/*.rb` in ruby/fiddle.

This is too brittle and as I already said it would break silently.
The next TruffleRuby release is in 6 months, I don't think it's a good idea to require no changes in Fiddle Ruby files for 6 months.
Also it doesn't solve it for JRuby.

---

@kou Let's try https://github.com/ruby/fiddle/pull/146#discussion_r1772483803, are you OK with that for now?

Reply:
I tried https://github.com/ruby/fiddle/pull/146#discussion_r1772483803 after `rake install` and with `ruby -ve 'require "fiddle"; puts $"; puts; require "fiddle"; puts $"'`, it seems to work fine.

I pushed a commit switching to that approach, I hope that addresses your concerns.

Reply:
Actually that approach is broken, as can be seen when trying to run the test suite, because `fiddle.rb` does not load all fiddle Ruby files but only some of them. For example `fiddle/pack.rb` isn't loaded eagerly.
And so when the user or tests does `require 'fiddle/struct'` we'll end up with a mix of files from stdlib and from the gem:
```
/home/eregon/.rubies/truffleruby-24.1.0/lib/truffle/truffle/fiddle_backend.rb
/home/eregon/.rubies/truffleruby-24.1.0/lib/mri/fiddle/closure.rb
/home/eregon/.rubies/truffleruby-24.1.0/lib/mri/fiddle/function.rb
/home/eregon/.rubies/truffleruby-24.1.0/lib/mri/fiddle/version.rb
/home/eregon/.rubies/truffleruby-24.1.0/lib/mri/fiddle.rb
/home/eregon/code/fiddle/lib/fiddle.rb
/home/eregon/code/fiddle/test/fiddle/helper.rb
/home/eregon/code/fiddle/lib/fiddle/value.rb
```

So I will remove the last commit.
`$LOAD_PATH.delete(__dir__)` is the only solution so far which works.

Reply:
> If this is a really short-term fix, we don't need `require_relative` changes. Because `lib/fiddle/**/*.rb` in ruby/fiddle and TruffleRuby will not be different. So existing TruffleRuby releases will work with `lib/fiddle/**/*.rb` in ruby/fiddle.

For completeness I tried and it doesn't work, the test suite fails during loading and just `ruby -Ilib -e 'require "fiddle"; require "fiddle/struct"'` fails with:
```
$ ruby -Ilib -e 'require "fiddle"; require "fiddle/struct"'
/home/eregon/code/fiddle/lib/fiddle/pack.rb:18:in `const_missing': uninitialized constant Fiddle::PackInfo::TYPE_BOOL (NameError)
Did you mean?  Fiddle::TYPE_LONG
               Fiddle::TYPE_VOID
	from /home/eregon/code/fiddle/lib/fiddle/pack.rb:18:in `<module:PackInfo>'
	from /home/eregon/code/fiddle/lib/fiddle/pack.rb:5:in `<module:Fiddle>'
	from /home/eregon/code/fiddle/lib/fiddle/pack.rb:4:in `<top (required)>'
	from <internal:core> core/kernel.rb:255:in `require'
	from /home/eregon/code/fiddle/lib/fiddle/struct.rb:4:in `<top (required)>'
	from <internal:core> core/kernel.rb:255:in `require'
	from -e:1:in `<main>'
```
OTOH `$LOAD_PATH.delete(__dir__)` works for that `ruby` command and the test suite loads and runs all tests (not all passing, but at least they all load).

Reply:
> > I said that I don't want to change any global variables by require "fiddle".
> 
> That wasn't clear to me, and I guess you mean change any global variable permanently, and that's why you prefer [#146 (comment)](https://github.com/ruby/fiddle/pull/146#discussion_r1772483803) to just `$LOAD_PATH.delete(__dir__)`, correct?

Ah, I understand. Correct.
 
> This is too brittle and as I already said it would break silently.

We can detect it by adding more checks to CI. For example, `ruby -e 'require "fiddle"; require "fiddle/struct"'` instead of `ruby -e 'require "fiddle"; require "fiddle/struct"'` as you showed.

> The next TruffleRuby release is in 6 months, I don't think it's a good idea to require no changes in Fiddle Ruby files for 6 months.

We can use TruffleRuby compatible change something like if we have a CI job for TruffleRuby:

```diff
diff --git a/lib/fiddle/pack.rb b/lib/fiddle/pack.rb
index 81088f4..99e0742 100644
--- a/lib/fiddle/pack.rb
+++ b/lib/fiddle/pack.rb
@@ -15,8 +15,11 @@ module Fiddle
       TYPE_USHORT => ALIGN_SHORT,
       TYPE_UINT   => ALIGN_INT,
       TYPE_ULONG  => ALIGN_LONG,
-      TYPE_BOOL   => ALIGN_BOOL,
     }
+    # TruffleRuby doesn't have Fiddle::TYPE_BOOL yet.
+    if Fiddle.const_defined?(:TYPE_BOOL)
+      ALIGN_MAP[TYPE_BOOL] = ALIGN_BOOL
+    end
 
     PACK_MAP = {
       TYPE_VOIDP => "L!",
@@ -31,15 +34,18 @@ module Fiddle
       TYPE_UINT   => "I!",
       TYPE_ULONG  => "L!",
     }
-    case SIZEOF_BOOL
-    when SIZEOF_CHAR
-      PACK_MAP[TYPE_BOOL] = PACK_MAP[TYPE_UCHAR]
-    when SIZEOF_SHORT
-      PACK_MAP[TYPE_BOOL] = PACK_MAP[TYPE_USHORT]
-    when SIZEOF_INT
-      PACK_MAP[TYPE_BOOL] = PACK_MAP[TYPE_UINT]
-    when SIZEOF_LONG
-      PACK_MAP[TYPE_BOOL] = PACK_MAP[TYPE_ULONG]
+    # TruffleRuby doesn't have Fiddle::SIZEOF_BOOL yet.
+    if Fiddle.const_defined?(:SIZEOF_BOOL)
+      case SIZEOF_BOOL
+      when SIZEOF_CHAR
+        PACK_MAP[TYPE_BOOL] = PACK_MAP[TYPE_UCHAR]
+      when SIZEOF_SHORT
+        PACK_MAP[TYPE_BOOL] = PACK_MAP[TYPE_USHORT]
+      when SIZEOF_INT
+        PACK_MAP[TYPE_BOOL] = PACK_MAP[TYPE_UINT]
+      when SIZEOF_LONG
+        PACK_MAP[TYPE_BOOL] = PACK_MAP[TYPE_ULONG]
+      end
     end
 
     SIZE_MAP = {
@@ -54,8 +60,11 @@ module Fiddle
       TYPE_USHORT => SIZEOF_SHORT,
       TYPE_UINT   => SIZEOF_INT,
       TYPE_ULONG  => SIZEOF_LONG,
-      TYPE_BOOL   => SIZEOF_BOOL,
     }
+    # TruffleRuby doesn't have Fiddle::TYPE_BOOL yet.
+    if Fiddle.const_defined?(:TYPE_BOOL)
+      SIZE_MAP[TYPE_BOOL] = SIZEOF_BOOL
+    end
     if defined?(TYPE_LONG_LONG)
       ALIGN_MAP[TYPE_LONG_LONG] = ALIGN_MAP[TYPE_ULONG_LONG] = ALIGN_LONG_LONG
       PACK_MAP[TYPE_LONG_LONG] = "q"
```

> Also it doesn't solve it for JRuby.

We can release an empty gem for JRuby like other gems did.

> Actually that approach is broken, as can be seen when trying to run the test suite, because `fiddle.rb` does not load all fiddle Ruby files but only some of them. For example `fiddle/pack.rb` isn't loaded eagerly. And so when the user or tests does `require 'fiddle/struct'` we'll end up with a mix of files from stdlib and from the gem:

How about requiring `fiddle/struct` in `fiddle.rb`?

```ruby
  stdlib = RbConfig::CONFIG["rubylibdir"]
  $LOAD_PATH.prepend(stdlib)
  begin
    require 'fiddle'
    require 'fiddle/struct'
  ensure
    if $LOAD_PATH.first == stdlib
      $LOAD_PATH.shift
    else
      raise 'fiddle files unexpectedly modified the $LOAD_PATH'
    end
  end
```

Or we can update `fiddle/pack.rb` in ruby/fiddle to TruffleRuby compatible as I showed.

Reply:
@kou I am not sure why you are trying to suggest even more complex or brittle solutions.
(brittle because `ruby -e 'require "fiddle"; require "fiddle/struct"'` is not proper testing and only covers one example, there are other fiddle files that are not eagerly loaded)
I have shown a solution that works and I have shown the various solutions your proposed so far do not work.
That solution is generic, i.e. it works for other gems too, not just fiddle, and we'll need to do something similar for some other default/bundled extension gems too.
So I have no interest to have highly-fiddle-specific solution here, it's just making things more complicated and less consistent.

Also the solution I propose is effectively equivalent to what is already done in `strscan` and `stringio`, where there it works because there is no `.rb` file, which is conceptually equivalent to not adding the gem `lib/` dir to `$LOAD_PATH`.

It is natural for language implementations to reimplement some of the core & standard library based on what makes sense for them, and possibly upstream that if it is stable enough and does not rely on too many internals. This upstreaming is a significant time investment, it does not happen in days but typically takes weeks or months (as we have already seen). It also significantly hinders the ability of language implementations to optimize these parts of stdlib better because then that upstreamed code needs to care about compatibility e.g. with older TruffleRuby versions, etc.

What is a concrete objective problem with `$LOAD_PATH.delete(__dir__)`?

It seems clear that "$LOAD_PATH is a global variable and shouldn't be changed" is not a good enough argument to reject that approach, `$LOAD_PATH` is part of the require API, it's the only way to influence where `require` looks. It is the natural way to solve this issue (short-term), where the gem is unexpectedly shadowing the standard library (which worked fine until now, as very few gems/Gemfile depended on `fiddle`).

> We can release an empty gem for JRuby like other gems did.

How is this any better? Am I missing something?
As far as I understand, it would require you to do more work by building + pushing + testing the extra java-platform gem (if forgotten it breaks things) and it's effectively exactly the same as `$LOAD_PATH.delete(__dir__)`.

And https://github.com/ruby/fiddle/issues/104 would also be more work for you as then you'd need to maintain a Java implementation.

This PR, once merged & released should require no work or effort from you at all. I thought you would like that.

Also to give you my point of view:
* CRuby broke some stdlib/default/bundled gems for other Ruby implementations by making them bundled gems (or adding warnings for them)
* I filed the issue in https://github.com/ruby/fiddle/issues/145
* @andrykonchin and I created a fix in this PR
* Now we are arguing for days about how it should be done, it is frustrating. It must be for you as well, I'm sorry for that.
* I don't have much time left to spend on this.

If there is still no progress on this PR, I guess one approach is to add some kind of patch in TruffleRuby to do `$LOAD_PATH.delete(fiddle_gem_lib_dir)` when `require 'fiddle'` to never load fiddle from the gem.
I don't think that's a good approach/solution though because:
* It won't fix existing TruffleRuby & JRuby releases.
* It's hidden vs being clear here in the gem.

Reply:
> I am not sure why you are suggesting more complex or brittle solutions.

I can't understand why I need to say the same thing again and again:
I don't want to change any global variables by require "fiddle".

I can't understand why you say that it's not clear. 

>  (brittle because `ruby -e 'require "fiddle"; require "fiddle/struct"'` is not proper testing and only covers one example, there are other fiddle files that are not eagerly loaded) I have shown a solution that works reliably and I have shown the various solutions your proposed so far do not work.

I already showed a solution for it. We can use it for other files.

> That solution is generic, i.e. it works for other gems too, not just fiddle, and we'll need to do something similar for some other default/bundled extension gems too. So I have no interest to have highly-fiddle-specific solution here, it's just making things more complicated and less consistent (but if it would work reliably, fine, I'd be OK with it).
>
> Also the solution I propose is effectively equivalent to what is already done in `strscan` and `stringio`, where there it works because there is no `.rb` file, which is conceptually equivalent to not adding the gem `lib/` dir to `$LOAD_PATH`.

In principal, your solution (ignore gem) isn't a good solution. If an user wants to use the specific version of A gem, the version information is just ignored with TruffleRuby. It just avoids `require` error. I pointed out this 2 years ago: https://github.com/ruby/strscan/pull/35#discussion_r856753083

It seems that you say that this is a "short-term" workaround but you don't think that this is a "short-term" workaround. At least strscan still uses this workaround. I don't think that it's "short-term".

Anyway, here are PRs that import JRuby and TruffleRuby implementations:

* GH-147
* GH-148

Reply:
> I don't want to change any global variables by require "fiddle".

But without that, there is AFAIK no short-term solution to make it work.
Hence while I understand you don't like to change `$LOAD_PATH`, since it is the only solution to fix existing releases (broken indirectly by the fiddle warning added in CRuby), I think it is a reasonable short-term solution.

> I already showed a solution for it. We can use it for other files.

It does feel brittle as if a new file is added it could be missed there. I suppose we could use `Dir.glob` but anyway that snippet of code is just more complicated and feels more risky than `$LOAD_PATH.delete(__dir__)`.
With `$LOAD_PATH.delete(__dir__)` the behavior is clear, with the alternatives it's far more subtle and therefore prone to bugs.

> In principal, your solution (ignore gem) isn't a good solution. If an user wants to use the specific version of A gem, the version information is just ignored with TruffleRuby. 

Yes that's a valid point.
However, a subset of the functionality working is much better than nothing working or even `require 'fiddle'` failing.
The tension here is with:
> It also significantly hinders the ability of language implementations to optimize these parts of stdlib better because then that upstreamed code needs to care about compatibility e.g. with older TruffleRuby versions, etc.

And for example we found that synchronization for `stringio` is a huge overhead, and optimizing that would be much harder if TruffleRuby's stringio implementation lives in ruby/stringio and needs to support multiple truffleruby versions.

> It seems that you say that this is a "short-term" workaround but you don't think that this is a "short-term" workaround. At least strscan still uses this workaround. I don't think that it's "short-term".

Actually on September 12 I wrote to @andrykonchin in a private chat:
> mmh strscan is adding some new features and they are not added for truffleruby since we have `strscan.rb` in the truffleruby repo: https://github.com/ruby/strscan/pull/106
> Probably we should consider moving that upstream. We need to look at the stability of the Primitive's used there though, and Truffle:: methods

And there is the same concern with moving the code here.

> Anyway, here are PRs that import JRuby and TruffleRuby implementations:

Thank you, I didn't expect that.
There are some concerns we need to discuss before that can be merged, I'll reply on that PR.